### ISSUE TITLE
GCS - signed URLs (local development)

### ISSUE THREAD

**Issue by jpaulgs** [googleapis/google-cloud-ruby#6268]
#### Environment details

  - OS: OSX 
  - Ruby version: 2.6.3
  - Gem name and version: google-cloud-storage 1.26.1

#### Steps to reproduce

  1. gcloud auth application-default login
  1. accept oauth flow in browser
  1. run the following code

I've also logged in using `gcloud auth login` with the same results.

#### Code example

Replace 'some_bucket_name' with a bucket you control
```ruby
# List files in a bucket 
irb(main):001:0> require 'google/cloud/storage'
irb(main):002:0> storage = Google::Cloud::Storage.new
irb(main):003:0> bucket = storage.bucket('some_bucket_name')
irb(main):004:0> bucket.files
=> []
# that has worked as the bucket I'm pointing to does not have any files
irb(main):005:0> bucket.signed_url('test.txt', method: 'PUT', expires: 600)
Traceback (most recent call last):
        6: from /Users/jeromepaul/.rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        5: from /Users/jeromepaul/.rbenv/versions/2.6.3/bin/irb:23:in `load'
        4: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.2.1/exe/irb:11:in `<top (required)>'
        3: from (irb):7
        2: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/google-cloud-storage-1.26.1/lib/google/cloud/storage/bucket.rb:1551:in `signed_url'
        1: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/google-cloud-storage-1.26.1/lib/google/cloud/storage/file/signer_v2.rb:122:in `signed_url'
Google::Cloud::Storage::SignedUrlUnavailable (Google::Cloud::Storage::SignedUrlUnavailable)
```

The documentation for application default credentials suggests that this should work: https://cloud.google.com/iam/docs/service-accounts#application_default_credentials

> The most common use case is testing code on a local machine, and then moving to a development project in Google Cloud, and then moving to a production project in Google Cloud. Using Application Default Credentials ensures that the service account works seamlessly; when testing on your local machine, it uses a locally-stored service account key, but when running on Compute Engine, it uses the project’s default Compute Engine service account.

I've also looked at https://googleapis.dev/ruby/google-cloud-storage/latest/file.AUTHENTICATION.html#cloud-sdk which again suggests that this should work for local development.

I know from the documentation that a `SignedUrlUnavailable` error is raised when 'when File#signed_url is unable to generate a URL due to missing credentials needed to create the URL.' However I can't find documentation that indicates how to create the required key.

Can the error class please be updated to point to relevant documentation?

**Comment by jpaulgs** [comment]
To be honest for now the documentation should really reference: https://cloud.google.com/storage/docs/reference/libraries#setting_up_authentication

**Comment by quartzmo** [comment]
@jpaulgs Thank you for this valuable feedback. We'll see what we can do to improve this. Much appreciated! I'll try to reproduce your code example as soon as I can.

**Comment by chvreddy** [comment]
@quartzmo Having the same issue in GKE with workload identity feature 

```rb(main):005:0> bucket.signed_url('test.txt', method: 'PUT', expires: 600)
Traceback (most recent call last):
        6: from /Users/jeromepaul/.rbenv/versions/2.6.3/bin/irb:23:in `<main>'
        5: from /Users/jeromepaul/.rbenv/versions/2.6.3/bin/irb:23:in `load'
        4: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/irb-1.2.1/exe/irb:11:in `<top (required)>'
        3: from (irb):7
        2: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/google-cloud-storage-1.26.1/lib/google/cloud/storage/bucket.rb:1551:in `signed_url'
        1: from /Users/jeromepaul/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/google-cloud-storage-1.26.1/lib/google/cloud/storage/file/signer_v2.rb:122:in `signed_url'
Google::Cloud::Storage::SignedUrlUnavailable (Google::Cloud::Storage::SignedUrlUnavailable)```

This currently relies on having a private key for the service account that is used to sign requests. It's not clear to me how this would be done while using Workload Identity.

**Comment by quartzmo** [comment]
@frankyn Any idea about using Signed URL V2 in GKE with Workload Identity?

**Comment by frankyn** [comment]
Hi @quartzmo, it would need to follow a similar solution for Compute Engine instances IIRC. I thought the Ruby library support IAM SignBlob for Signed URLs*

**Comment by quartzmo** [comment]
> I thought the Ruby library support IAM SignBlob for Signed URLs*

@frankyn You mean this [signBlob](https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/signBlob) RPC? I don't remember using it in google-cloud-storage; am I forgetting something? (Also, the docs say that method is deprecated.)

**Comment by chvreddy** [comment]
@quartzmo @frankyn 
tried implementing  signBlob but got an error when I ran 

response = service.sign_service_account_blob(name, request_body)

Google::Apis::ClientError: forbidden: Permission iam.serviceAccounts.signBlob is required to perform this operation on service account projects/{project-id}/serviceAccounts/{SA-Email}

But my SA account has Service Account Token Creator
So i am curious what went wrong, can we use signBlob with workload identity or its permission issue from my side.

**Comment by chvreddy** [comment]
@quartzmo @frankyn

Implemented same in google cloud shell
but got different error

`Google::Apis::ClientError: badRequest: Request contains an invalid argument.'

Followed link: https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/signBlob

So i am curious what went wrong, can we use signBlob with workload identity

**Comment by quartzmo** [comment]
@jpaulgs wrote:

> I know from the documentation that a SignedUrlUnavailable error is raised when 'when File#signed_url is unable to generate a URL due to missing credentials needed to create the URL.' However I can't find documentation that indicates how to create the required key.

Regarding this error, please see the answer provided by @dazuma on [this Stack Overflow question](https://stackoverflow.com/questions/50549176/google-cloud-storage-500-internal-server-error-googlecloudstoragesignedur):

> Google App Engine (as well as Google Compute Engine, Kubernetes Engine, and Cloud Run) provides "ambient" credentials associated with the VM or instance being run, but only in the form of OAuth tokens. For most API calls, this is sufficient and convenient.

> However, there are a small number of exceptions, and Google Cloud Storage is one of them. Recent Storage clients (including the google-cloud-storage gem) may require a full service account key to support certain calls that involve signed URLs. This full key is not provided automatically by App Engine (or other hosting environments). You need to provide one yourself. So as a previous answer indicated, if you're using Cloud Storage, you may not be able to depend on the "ambient" credentials. Instead, you should create a service account, download a service account key, and make it available to your app (for example, via the ActiveStorage configs, or by setting the `GOOGLE_APPLICATION_CREDENTIALS` environment variable).

**Comment by quartzmo** [comment]
@jpaulgs asked:

> Can the error class please be updated to point to relevant documentation?

I will open a PR to improve the error detail messages for all `SignedUrlUnavailable` errors, and also update the documentation  for the three public `#signed_url` methods.

**Comment by quartzmo** [comment]
Here is the new error message:

```
irb(main):003:0> bucket.signed_url('test.txt', method: 'PUT', expires: 600)
Traceback (most recent call last):

...

Google::Cloud::Storage::SignedUrlUnavailable (Service account credentials 'issuer (client_email)' is missing. To generate service account credentials see https://cloud.google.com/storage/docs/authentication#service_accounts)
```

**Comment by quartzmo** [comment]
@chvreddy I do not believe that you can use [Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) with the current signed URLs feature. If you find a way to do it, please post the solution here. Thank you!

**Comment by frankyn** [comment]
Hi folks, reopening.

I missed follow-ups on this thread, apologies.

There are two issues:

1. Ruby Storage library doesn't have an easy way to pass in a signer to introduce an IAM API call [signBlob](https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/signBlob). 
1. (still a little fuzzy) You will need to grant role `roles/iam.serviceAccountTokenCreator` to the Google Service account binded to your `project-id.svc.id.goog[k8s-namespace/ksa-name]`, e.g. `workload-service-account@project-id.iam.gserviceaccount.com`.This same Google Service account email must be passed into the signed_url method as well using `issuer`.

I provided a prototype that I used to make a successful request within a GKE cluster using Work Identity. I'm new to this environment but hopefully this can help move the conversation forward.

@quartzmo I created a class IAMSigner to get past using an RSA object, and it would be helpful to have a way to provide a lambda that's called instead of the RSA sign method. 

```ruby
require 'googleauth'
require 'google/cloud/storage'
require 'google/apis/iamcredentials_v1'
require 'base64'

IAMCredentials = Google::Apis::IamcredentialsV1 # Alias the module


# The following is a hack to introduce IAMCredentialsService into the Storage library
# without modifying the existing library as a proof of concept.
class IAMSigner
   def initialize(issuer)
     @iam_credentials_client = IAMCredentials::IAMCredentialsService.new
     # Get the environment configured authorization
     scopes =  ['https://www.googleapis.com/auth/cloud-platform',
                'https://www.googleapis.com/auth/compute']
     @iam_credentials_client.authorization = Google::Auth.get_application_default(scopes)
     @issuer = issuer
   end

   def sign(digest, string_to_sign)
      # Ignore digest
      request = {
            "payload": string_to_sign,
      }
      response = @iam_credentials_client.sign_service_account_blob(
        "projects/-/serviceAccounts/#{@issuer}",
        request,
        {}
      )
      response.signed_blob
   end
end

storage = Google::Cloud::Storage.new
storage_expiry_time = 5 * 60 # 5 minutes
bucket_name = "anima-frank"
file_name = "unnamed.jpg"
# Digging into this a bit more.
issuer = "workload-identity-service-account"
signing_key = IAMSigner.new issuer

url = storage.signed_url bucket_name, file_name, issuer: issuer,
                         signing_key: signing_key,
                         method: "GET", expires: storage_expiry_time,
                         version: :v4

puts url
````

**Comment by quartzmo** [comment]
In the last example, should it be `issuer_and_signer`, given the order of the array `[issuer, response.signed_blob]` ?

**Comment by frankyn** [comment]
Completed example doesn't consider issuer being needed to construct string_to_sign. 

Let's follow Go's interface a little bit instead with https://github.com/googleapis/google-cloud-go/issues/1130#issuecomment-484236791

So updated the example, it would be:

```ruby
require 'googleauth'
require 'google/cloud/storage'
require 'google/apis/iamcredentials_v1'

IAMCredentials = Google::Apis::IamcredentialsV1 # Alias the module

iam_credentials_client = IAMCredentials::IAMCredentialsService.new
# Get the environment configured authorization
scopes =  ['https://www.googleapis.com/auth/cloud-platform',
            'https://www.googleapis.com/auth/compute']
iam_credentials_client.authorization = Google::Auth.get_application_default(scopes)     
issuer =  iam_credentials_client.authorization.issuer

signer = lambda do |string_to_sign|
     request = {
           "payload": string_to_sign,
     }
     response = iam_credentials_client.sign_service_account_blob(
       "projects/-/serviceAccounts/#{issuer}",
       request,
       {}
     )
     response.signed_blob
end

storage = Google::Cloud::Storage.new
storage_expiry_time = 5 * 60 # 5 minutes
bucket_name = "anima-frank"
file_name = "unnamed.jpg"

url = storage.signed_url bucket_name, file_name, issuer: issuer, signer: signer
                         method: "GET", expires: storage_expiry_time,
                         version: :v4

puts url

```

**Comment by quartzmo** [comment]
So the only public API or behavior change needed in this last example is that `signing_key` (or its alias `private_key`) can be a lambda which when called with `string_to_sign`, returns the `signature`?

**Comment by frankyn** [comment]
That's correct @quartzmo, I think maybe a new alias such as `signer` may help as well but using `signing_key` is a good starting point.

**Comment by frankyn** [comment]
@chvreddy could you confirm if https://github.com/googleapis/google-cloud-ruby/issues/6268#issuecomment-663261941 would help in this case?

**Comment by chvreddy** [comment]
Thanks @frankyn let me apply this patch and see if we are able to create signed URL's

**Comment by frankyn** [comment]
Thanks for confirming @chvreddy.
If you're attempting to use one of these, I recommend looking at the following example:
**(Warning it's still a prototype)**
```ruby
require 'googleauth'
require 'google/cloud/storage'
require 'google/apis/iamcredentials_v1'

IAMCredentials = Google::Apis::IamcredentialsV1 # Alias the module


# The following is a hack to introduce IAMCredentialsService into the Storage library
# without modifying the existing library as a proof of concept.
class IAMSigner
   def initialize(issuer)
     @iam_credentials_client = IAMCredentials::IAMCredentialsService.new
     # Get the environment configured authorization
     scopes =  ['https://www.googleapis.com/auth/cloud-platform',
                'https://www.googleapis.com/auth/compute']
     @iam_credentials_client.authorization = Google::Auth.get_application_default(scopes)
     @issuer = issuer
   end

   def sign(digest, string_to_sign)
      # Ignore digest
      request = {
            "payload": string_to_sign,
      }
      response = @iam_credentials_client.sign_service_account_blob(
        "projects/-/serviceAccounts/#{@issuer}",
        request,
        {}
      )
      response.signed_blob
   end
end

storage = Google::Cloud::Storage.new
storage_expiry_time = 5 * 60 # 5 minutes
bucket_name = "anima-frank"
file_name = "unnamed.jpg"
issuer = "workload-service-account@project-id.iam.gserviceaccount.com"
signing_key = IAMSigner.new issuer

url = storage.signed_url bucket_name, file_name, issuer: issuer,
                         signing_key: signing_key,
                         method: "GET", expires: storage_expiry_time,
                         version: :v4

puts url
```

**Comment by chvreddy** [comment]
> Thanks for confirming @chvreddy.
> If you're attempting to use one of these, I recommend looking at the following example:
> **(Warning it's still a prototype)**
> 
> ```ruby
> require 'googleauth'
> require 'google/cloud/storage'
> require 'google/apis/iamcredentials_v1'
> 
> IAMCredentials = Google::Apis::IamcredentialsV1 # Alias the module
> 
> 
> # The following is a hack to introduce IAMCredentialsService into the Storage library
> # without modifying the existing library as a proof of concept.
> class IAMSigner
>    def initialize(issuer)
>      @iam_credentials_client = IAMCredentials::IAMCredentialsService.new
>      # Get the environment configured authorization
>      scopes =  ['https://www.googleapis.com/auth/cloud-platform',
>                 'https://www.googleapis.com/auth/compute']
>      @iam_credentials_client.authorization = Google::Auth.get_application_default(scopes)
>      @issuer = issuer
>    end
> 
>    def sign(digest, string_to_sign)
>       # Ignore digest
>       request = {
>             "payload": string_to_sign,
>       }
>       response = @iam_credentials_client.sign_service_account_blob(
>         "projects/-/serviceAccounts/#{@issuer}",
>         request,
>         {}
>       )
>       response.signed_blob
>    end
> end
> 
> storage = Google::Cloud::Storage.new
> storage_expiry_time = 5 * 60 # 5 minutes
> bucket_name = "anima-frank"
> file_name = "unnamed.jpg"
> issuer = "workload-service-account@project-id.iam.gserviceaccount.com"
> signing_key = IAMSigner.new issuer
> 
> url = storage.signed_url bucket_name, file_name, issuer: issuer,
>                          signing_key: signing_key,
>                          method: "GET", expires: storage_expiry_time,
>                          version: :v4
> 
> puts url
> ```

sure @frankyn  will apply this patch and see how is it going. Thanks!

**Comment by frankyn** [comment]
Update add in #7091 is pending release of google-cloud-storage gem. 

reopening until gem is released.

**Comment by joedicator** [comment]
@frankyn 🆒 Are you able to comment when gem will be released?

**Comment by frankyn** [comment]
Hi @joedicator,

I'm pending input from @quartzmo right now. Open PR with next version is https://github.com/googleapis/google-cloud-ruby/pull/6993

**Comment by frankyn** [comment]
Short update: @quartzmo is aiming to release the library tomorrow. 

Thanks @quartzmo!

**Comment by frankyn** [comment]
Woot! Thanks @quartzmo!

[google-cloud-storage 1.27.0 released](https://rubygems.org/gems/google-cloud-storage/versions/1.27.0)
[Updated API documentation](https://googleapis.dev/ruby/google-cloud-storage/v1.27.0/Google/Cloud/Storage/Bucket.html#signed_url-instance_method)

Closing this issue, please reopen if there's still open questions. Thank you for your patience on this issue.

**Comment by quartzmo** [comment]
Great addition to google-cloud-storage, thank you @frankyn!

**Comment by quartzmo** [comment]
Closed by #7091

**Comment by chvreddy** [comment]
Thanks @quartzmo @frankyn for the support
@frankyn the monkey patch you provided worked great, Thanks! and @frankyn thanks for providing the upgraded gem quickly.

**Comment by frankyn** [comment]
Thanks @chvreddy for letting us know, it's really great to hear that you were unblocked!

**Comment by stefanahman** [comment]
How can Workload Identity be used in conjunction with [ActiveStorage - Google Cloud Storage](https://edgeguides.rubyonrails.org/active_storage_overview.html#google-cloud-storage-service)?

**Comment by frankyn** [comment]
Hi @stefanahman, thanks for raising your question!
 
Could you start a new issue with more background on your use case to better support you?

**Comment by mroach** [comment]
@stefanahman Did you end up creating an issue and/or figuring it out? 

I implemented this signing code and that works, but now I have an implementation dilemma since Google Cloud Storage is only used in production, not tests or development.

**Comment by stefanahman** [comment]
> @stefanahman Did you end up creating an issue and/or figuring it out?
> 
> I implemented this signing code and that works, but now I have an implementation dilemma since Google Cloud Storage is only used in production, not tests or development.

No, sorry. I haven't looked into it more.
But you specify the storage per environment, right? Just specify "Disk" as service for those environments:

```yml
test:
  service: Disk
  root: <%= Rails.root.join("tmp/storage") %>

local:
  service: Disk
  root: <%= Rails.root.join("storage") %>
```

**Comment by sephethus** [comment]
This issue still persists in 2023, workload identity is now the prevailing solution in GKE and likely other cloud based kubernetes solutions. What is the status on this?

**Comment by palladius** [comment]
Any update here? It seems most of the effort has been done.
@dazuma can you maybe pick this up?
